### PR TITLE
Added support of custom matchers in static mappings

### PR DESCRIPTION
**Pull request by levanoz**

### CODE REVIEW

File: src/WireMock.Net/Serialization/MatcherMapper.cs [hunk 1]
```diff
@@ -105,6 +105,9 @@ public IMatcher Map([CanBeNull] MatcherModel matcher)
                     return new SimMetricsMatcher(matchBehaviour, stringPatterns, type, throwExceptionWhenMatcherFails);
 
                 default:
+                    if (_settings.CustomMatcherMapping != null && _settings.CustomMatcherMapping.ContainsKey(matcherName))
+                        return _settings.CustomMatcherMapping[matcherName](matcher);
```
Comment (StefH) [inline src/WireMock.Net/Serialization/MatcherMapper.cs:109]: Add `{` and `}`.

File: src/WireMock.Net/Settings/IWireMockServerSettings.cs [hunk 2]
```diff
@@ -229,5 +231,11 @@ public interface IWireMockServerSettings
         /// </summary>
         [PublicAPI]
         bool? SaveUnmatchedRequests { get; set; }
+
+        /// <summary>
+        /// Custom matcher mappings for static mappings
+        /// </summary>
+        [PublicAPI]
+        IDictionary<string, Func<MatcherModel, IMatcher>> CustomMatcherMapping { get; set; }
```
Comment (StefH) [inline src/WireMock.Net/Settings/IWireMockServerSettings.cs:239]: CustomMatcherMappings

File: test/WireMock.Net.Tests/Serialization/CustomPathParamMatcher.cs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            _path = path;
+            _pathParts = GetPathParts(path);
+            _pathParams = pathParams.ToDictionary(x => x.Key, x => x.Value, StringComparer.OrdinalIgnoreCase);
+        }
+
+        public double IsMatch(string input)
+        {
+            var inputParts = GetPathParts(input);
+            if (inputParts.Length != _pathParts.Length)
+                return MatchScores.Mismatch;
```
Comment (StefH) [inline test/WireMock.Net.Tests/Serialization/CustomPathParamMatcher.cs:42]: Add `{` and `}`.

**Review by StefH — CHANGES_REQUESTED** [review]
Thank you for this PR.

I think all comments are resolved, but can you double check?

File: test/WireMock.Net.Tests/Serialization/CustomPathParamMatcher.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                    throw;
+            }
+
+            return MatchScores.Perfect;
+        }
+
+        public AnyOf<string, StringPattern>[] GetPatterns()
+        {
+            return new[] { new AnyOf<string, StringPattern>(JsonConvert.SerializeObject(new CustomPathParamMatcherModel(_path, _pathParams))) };
+            //return _pathParams.Values.Select(x => new AnyOf<string, StringPattern>(x)).ToArray();
```
[the remark this answers is not in the document]
Comment (levanoz) [reply test/WireMock.Net.Tests/Serialization/CustomPathParamMatcher.cs]: Yes, I've removed it in latest commit

File: test/WireMock.Net.Tests/Serialization/CustomPathParamMatcher.cs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        }
+
+        private string[] GetPathParts(string path)
+        {
+            var hashMarkIndex = path.IndexOf('#');
+            if (hashMarkIndex != -1)
+                path = path.Substring(0, hashMarkIndex);
```
[the remark this answers is not in the document]
Comment (levanoz) [reply test/WireMock.Net.Tests/Serialization/CustomPathParamMatcher.cs:95]: I've added them in latest commit

File: test/WireMock.Net.Tests/Serialization/CustomPathParamMatcher.cs [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+            var queryParamsIndex = path.IndexOf('?');
+            if (queryParamsIndex != -1)
+                path = path.Substring(0, queryParamsIndex);
```
[the remark this answers is not in the document]
Comment (levanoz) [reply test/WireMock.Net.Tests/Serialization/CustomPathParamMatcher.cs:101]: I've added them in latest commit

File: test/WireMock.Net.Tests/Serialization/CustomPathParamMatcher.cs [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+            {
+                for (int i = 0; i < inputParts.Length; i++)
+                {
+                    var inputPart = inputParts[i];
+                    var pathPart = _pathParts[i];
+                    if (pathPart.StartsWith("{") && pathPart.EndsWith("}"))
+                    {
+                        var pathParamName = pathPart.Trim('{').Trim('}');
+                        if (!_pathParams.ContainsKey(pathParamName))
+                            return MatchScores.Mismatch;
```
[the remark this answers is not in the document]
Comment (levanoz) [reply test/WireMock.Net.Tests/Serialization/CustomPathParamMatcher.cs:57]: I've added them in latest commit

File: test/WireMock.Net.Tests/Serialization/CustomPathParamMatcher.cs [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+                        if (!Regex.IsMatch(inputPart, _pathParams[pathParamName], RegexOptions.IgnoreCase))
+                            return MatchScores.Mismatch;
```
[the remark this answers is not in the document]
Comment (levanoz) [reply test/WireMock.Net.Tests/Serialization/CustomPathParamMatcher.cs:62]: I've added them in latest commit

File: test/WireMock.Net.Tests/Serialization/CustomPathParamMatcher.cs [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+                    }
+                    else
+                    {
+                        if (!inputPart.Equals(pathPart, StringComparison.InvariantCultureIgnoreCase))
+                            return MatchScores.Mismatch;
```
[the remark this answers is not in the document]
Comment (levanoz) [reply test/WireMock.Net.Tests/Serialization/CustomPathParamMatcher.cs:69]: I've added them in latest commit

File: test/WireMock.Net.Tests/Serialization/CustomPathParamMatcher.cs [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+                    }
+                }
+            }
+            catch
+            {
+                if (ThrowException)
+                    throw;
```
[the remark this answers is not in the document]
Comment (levanoz) [reply test/WireMock.Net.Tests/Serialization/CustomPathParamMatcher.cs:78]: I've added them in latest commit

File: test/WireMock.Net.Tests/Serialization/MatcherModelMapperTests.cs [hunk 11]
```diff
@@ -310,5 +313,70 @@ public void MatcherModelMapper_Map_SimMetricsMatcher_Throws2()
...
+
+            // Act
+            var model = _sut.Map(matcher);
+
+            // Assert
+            using (new AssertionScope())
+            {
+                model.Should().NotBeNull();
+                model.Name.Should().Be(nameof(CustomPathParamMatcher));
+                var matcherParams = JsonConvert.DeserializeObject<CustomPathParamMatcherModel>((string)model.Pattern);
```
[the remark this answers is not in the document]
Comment (levanoz) [reply test/WireMock.Net.Tests/Serialization/MatcherModelMapperTests.cs:373]: Added empty line

**Comment by StefH** [comment]
@levanoz 
Thanks a lot.

Did you test this code change in a real environment?

**Comment by levanoz** [comment]
> @levanoz Thanks a lot.
> 
> Did you test this code change in a real environment?

@StefH 
My bad, I forgot to run it :( 
I just wrote unit tests for MatcherModelMapper and thought that would be enaugh.
Now I've run it, there was JSON serialization issue, which I've fixed by adding JsonIgnore attribute to CustomMatcherMappings property in settings. Is it okay?
Thanks a lot for reviews.

**Comment by levanoz** [comment]
> @levanoz I think it's correct for now.
> 
> I was thinking...
> 
> Maybe can you also add a 'real' test like this one:
> 
> https://github.com/WireMock-Net/WireMock.Net/blob/60bdc06d2979a978d2788ef68d300ec22f455960/test/WireMock.Net.Tests/WireMockServerTests.WithBody.cs#L49
> 
> But intead using the wildcardMatcher, use your own test-custom matcher.

@StefH 
I've added two 'real' tests:
https://github.com/WireMock-Net/WireMock.Net/blob/629da5aa2451b837116321a2f286194d25747b74/test/WireMock.Net.Tests/WireMockServerTests.cs#L394
https://github.com/WireMock-Net/WireMock.Net/blob/629da5aa2451b837116321a2f286194d25747b74/test/WireMock.Net.Tests/WireMockServerTests.cs#L442

**Review by StefH — APPROVED** [review]

**Comment by StefH** [comment]
Later today I'll merge and release a new version.

**Comment by levanoz** [comment]
@StefH 
Sounds great, thank you!

**Pull request merged**